Differentiate stats for the same blobstore operation with purposes

ywangd opened the pull request:
Today blobstore stats are collected against each HTTP operation, e.g. Get, List. This is not granular enough because the same HTTP operration can be performed for different purposes, e.g. cluster state, indices or translog. This PR adds a new Purpose enum to provide further breakdown for the same HTTP operation. 

Relates: ES-6800

ywangd committed:
wip

ywangd committed:
Differentiate between purposes for the same blobstore operation

DaveCTurner commented on modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java:
```diff
@@ -79,11 +81,14 @@ class S3BlobContainer extends AbstractBlobContainer {
 
     private final S3BlobStore blobStore;
     private final String keyPath;
+    @Nullable
+    private final Purpose purpose;
```
I would much prefer this not to be nullable, let's represent the cases that would be null as a proper enum value instead. Indeed at some point in the future I think we'll want to [distinguish metadata and data operations](https://github.com/elastic/elasticsearch/issues/81352) too, and this seems like a good mechanism for that.

ywangd replied:
Thanks David. I replaced `null` with `Purpose.Generic` in 0a399421c41e2cc96f0410e6a58414fb0707a249

DaveCTurner commented on server/src/main/java/org/elasticsearch/common/blobstore/BlobStore.java:
```diff
@@ -29,6 +32,14 @@ public interface BlobStore extends Closeable {
      */
     void deleteBlobsIgnoringIfNotExists(Iterator<String> blobNames) throws IOException;
 
+    default void deleteBlobsIgnoringIfNotExists(Iterator<String> blobNames, @Nullable Purpose purpose) throws IOException {
```
I would also much prefer not to have an implicit default for the `purpose` parameter like this. Experience shows that we will forget to set it in future work unless the caller is required to make an explicit choice.

That said, temporarily adding an implicit default is a reasonable strategy for avoiding breaking the serverless build, but it must only be temporary. Ideally we'd mark the overloads that don't take the `Purpose` as `@Deprecated(forRemoval = true)`.

ywangd replied:
Agreed especially since this interface method is pretty new. I will go through the deprecation and removal process once we are happy with the proposed approach in this PR.

ywangd committed:
Address feedback to drop nullable

ywangd committed:
address feedback and add deprecation

ywangd committed:
fix test

ywangd committed:
tweak and add tests

elasticsearchmachine commented:
Hi @ywangd, I've created a changelog YAML for you.

ywangd committed:
Update docs/changelog/99615.yaml

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

ywangd commented on server/src/main/java/org/elasticsearch/common/blobstore/BlobStore.java:
```diff
@@ -27,8 +29,25 @@ public interface BlobStore extends Closeable {
      * Delete all the provided blobs from the blob store. Each blob could belong to a different {@code BlobContainer}
      * @param blobNames the blobs to be deleted
      */
+    @Deprecated(forRemoval = true)
     void deleteBlobsIgnoringIfNotExists(Iterator<String> blobNames) throws IOException;
 
+    // TODO: Remove the default implementation and require each blob store to implement this method. Once it's done, remove the
+    // the above overload version that does not take the Purpose parameter.
+    /**
+     * Delete all the provided blobs from the blob store. Each blob could belong to a different {@code BlobContainer}
+     * @param blobNames the blobs to be deleted
+     * @param purpose the purpose of the {@code BlobContainer} associated to the blobs to be deleted. It should be set
+     *                to {@code Purpose.GENERIC}, if the blobs are from multiple {@code BlobContainer}s.
+     */
+    default void deleteBlobsIgnoringIfNotExists(Iterator<String> blobNames, Purpose purpose) throws IOException {
+        if (purpose == Purpose.GENERIC) {
+            deleteBlobsIgnoringIfNotExists(blobNames);
+        } else {
+            throw new UnsupportedOperationException();
+        }
+    }
```
As David [suggested](https://github.com/elastic/elasticsearch/pull/99615#discussion_r1328331423), I will perform the cleanup of these two methods in a follow-up and avoid breaking builds on either side.

ywangd commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into s3-more-granular-metrics

DaveCTurner commented on server/src/main/java/org/elasticsearch/common/blobstore/BlobPath.java:
```diff
@@ -82,4 +97,21 @@ public boolean equals(Object o) {
     }
+
+    public enum Purpose {
+        GENERIC("Generic"),
```
Could we call this `SNAPSHOT` instead of `GENERIC`? That way if we have another purpose in future we will have to add another entry here, because none of the existing entries will fit.

DaveCTurner commented on modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java:
```diff
@@ -326,29 +298,83 @@ ThreadPool getThreadPool() {
+            return new IgnoreNoResponseMetricsCollector() {
+                @Override
+                public void collectMetrics(Request<?> request) {
+                    if (Assertions.ENABLED) {
+                        // assertions based on operation
+                        switch (operation) {
+                            case GET_OBJECT, LIST_OBJECTS -> {
+                                assert request.getHttpMethod().name().equals("GET");
+                            }
+                            case PUT_OBJECT -> {
+                                assert request.getHttpMethod().name().equals("PUT");
+                            }
+                            case PUT_MULTIPART_OBJECT -> {
+                                assert request.getHttpMethod().name().equals("PUT") || request.getHttpMethod().name().equals("POST");
+                            }
+                            case DELETE_OBJECTS -> {
+                                assert request.getHttpMethod().name().equals("POST");
+                            }
+                            case ABORT_MULTIPART_OBJECT -> {
+                                assert request.getHttpMethod().name().equals("DELETE");
+                            }
+                            default -> throw new AssertionError("unknown operation category " + operation);
+                        }
+                    }
```
Nit: I'd rather things like this were a separate `boolean` method, called as `assert assertConsistentMethod(operation, request);`, instead of an explicit check for `Assertions.ENABLED`.

ywangd replied:
Yep updated as suggested 8eada5ddb82fb8d78c682fcf0ee9dc14ac3f323e

DaveCTurner reviewed (commented):
I think attaching the `Purpose` to the path to the container will be ok for our current needs, but I worry that it won't be flexible enough for some related future work: we want to distinguish data from metadata operations when taking a snapshot, and I can see value in being able to distinguish searchable-snapshot-related read operations from regular snapshot reads too.

Have we considered making this a per-operation property rather than a property of the `BlobContainer`? This will involve changing more call sites here of course but no more real complexity and a great deal more flexibility in future.

idegtiarenko commented on server/src/main/java/org/elasticsearch/common/blobstore/BlobPath.java:
```diff
@@ -26,16 +26,31 @@ public class BlobPath {
+    }
+
+    public BlobPath purpose(Purpose purpose) {
+        return new BlobPath(List.copyOf(this.paths), purpose);
```
Is it possible to change the purpose of current path?
If it is not the case, is it worth combining adding a path and setting the purpose into one call with something like `.add("/dir", purpose)` ?

ywangd replied:
Since we are heading to a different direction, the change here is no longer applicable.

henningandersen commented:
> Have we considered making this a per-operation property rather than a property of the BlobContainer? This will involve changing more call sites here of course but no more real complexity and a great deal more flexibility in future.

I agree, I'd like to at least try out that approach too to see how big of a change it is.

ywangd commented:
Thanks for the feedback. I was under the impression that once a `BlobContainer` is created, it can be used for only a single purpose since it deals with a flat list of files and no nested sub-folders. But apparently this was an invalid assumption. We can store both data and metadata under the same blobPath, e.g. shard level metadata and data files. I will take another attempt to make the change per operation.

ywangd commented:
Assuming "per operation" means "per method of the BlobContainer" class, I think it is achievable without having to change signature of every method. We still attach `purpose` to `BlobContainer` so that it is easily referenceable in all methods without having to change method signatures. Add a new `BlobContainer.withPurpose(Purpose)` method that returns a new `BlobContainer` with the given purpose. At each call site, we call `withPurpose` first when necessary and subsequent operation methods can remain unchanged. So it is roughly something like the follows:

```java
class BlobContainer {
    Purpose purpose = Purpose.SNAPSHOT;  // default

    BlobContainer withPurpose(Purpose purpose) {
        return new BlobContainer(..., purpose);
    }

    // operation method
    void writeBlob(...) {
        // the method can reference purpose as instance variable
    }

    ...
}

// Sample call site usage
class TranslogFileUploadTask {
    void doRun() {
        blobContainer.withPurpose(Purpose.TRANSLOG).writeBlob(...);
    }
}
```

I think the above can achieve the "per operation" variety with likely less cascading code changes. What do you think?

DaveCTurner commented:
> Add a new BlobContainer.withPurpose(Purpose) method that returns a new BlobContainer with the given purpose.

I'd rather not create a new `BlobContainer` for each method as would happen if we did that. And this approach would effectively make the purpose optional, which isn't what we want. It'd be better in the long run to just add a parameter throughout.

ywangd committed:
Address feedback to add the new purpose parameter to each operation method

ywangd committed:
tweak

ywangd commented on server/src/main/java/org/elasticsearch/common/blobstore/BlobContainer.java:
```diff
@@ -37,39 +37,55 @@ public interface BlobContainer {
-    InputStream readBlob(String blobName) throws IOException;
+    InputStream readBlob(OperationPurpose purpose, String blobName) throws IOException;
+
+    @Deprecated(forRemoval = true)
+    default InputStream readBlob(String blobName) throws IOException {
+        return readBlob(OperationPurpose.SNAPSHOT, blobName);
+    }
```
Similarly, methods marked as deprecated in this class will be removed once the necessary changes are made in the other repo. They are here to avoid breaking builds for now.

ywangd committed:
address feedback

ywangd commented on modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java:
```diff
@@ -326,29 +298,83 @@ ThreadPool getThreadPool() {
+        Map<String, Long> statsMap() {
+            return collectors.entrySet()
+                .stream()
+                .collect(Collectors.toUnmodifiableMap(Map.Entry::getKey, entry -> entry.getValue().counter.get()));
```
Thanks. I changed it to use `Maps.transformValues` since the result map has a different value type. https://github.com/elastic/elasticsearch/pull/99615/commits/29458a9f9645cd6877a73a91f1be80d6945723e2

ywangd commented on server/src/main/java/org/elasticsearch/common/blobstore/OperationPurpose.java:
```diff
@@ -0,0 +1,26 @@
+
+package org.elasticsearch.common.blobstore;
+
+public enum OperationPurpose {
+    SNAPSHOT("Snapshot"),
+    CLUSTER_STATE("ClusterState"),
+    INDICES("Indices"),
+    TRANSLOG("Translog");
```
These are the current set of operation purposes. I think we can start small with these ones. But I am also open to suggestions.

henningandersen replied:
I wonder about the name here, which hints that a blob could be used for different purposes, with different operations. I'd be inclined to assume that a blob has a single such purpose, unless we have good examples needing a per operation purpose?

I'd suggest `BlobPurpose` instead.

ywangd replied:
I was under the impression that we want to differentiate between operations (e.g. Get, Put) for the same purpose (e.g. ClusterState) since we both read and write to the cluster state (heartbeat?) blob. So my mental model for the relationship between operation and purpose was combinatory. I could be wrong. But do we want to answer questions like "how much does it cost to maintain translog"? If so, it seems to suggest that we do need to further differentiate "translog" for different operations?

henningandersen replied:
I agree that it is combinatoric. I see it as an extra stats field like
```
method="PUT" purpose="translog" count=7
```

but I'd expect a GET against the same file to use the same purpose. With the name `OperationPurpose`, it could be expected to see the same file being accessed with different purposes. I get that there could be some cases of list being indeterminate purpose though (and thus here see a separate one). I find that ok.

> So I updated accordingly.

Thanks.

DaveCTurner replied:
I think we may indeed want to track different operation purposes for individual blobs. For instance we may read a regular snapshot blob for a restore, or for a searchable snapshot (maybe distinguishing prewarming from cache misses) and maybe in future for integrity verification too.

henningandersen replied:
Ah, those purposes make sense to me too. I am good with the original name then.

ywangd replied:
David's example is much better and makes a lot sense to me. This confirms that we want to have the combinatory relationship between operation and purpose. It should guide us for the actual structure changes to stats collection which is left out for this PR and will be a follow-up immediately after this one.

For the class name, I kept it to be `BlobPurpose` since I feel it is more generic and can cover different intentions of purpose (be it operational or nominal).

idegtiarenko replied:
Should this be a class/record instead? We only use SNAPSHOT at the moment and the rest is used in plugins. Should it be declared there as well?

ywangd replied:
This Enum does have potential to evolve into an class/interface and offer more functionality. See also my comment [here](https://github.com/elastic/elasticsearch/pull/100236#issuecomment-1746165539). I think we can have follow-ups for it once we have more concrete ideas about usages. We originally wanted it for cost estimation, but then find it can be potentially be used for storage class, priority. We will need consilidate these and maybe more needs.

ywangd commented:
The PR is now good for another round. It is updated to have the new purpose parameter passed in for every operation method. The changeset is much larger, though most of them are rather mechanical. Also note the new parameter is handled only by the S3 implementation in this PR. Handling by other implementations can be added in follow-up PRs.

ywangd committed:
assertion method

ywangd committed:
fix tests

ywangd commented:
A gentle reminder for reviews, thanks!

DaveCTurner commented on modules/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java:
```diff
@@ -488,17 +489,18 @@ private void writeBlobMultipart(BlobInfo blobInfo, byte[] buffer, int offset, in
     /**
      * Deletes the given path and all its children.
      *
+     * @param purpose
```
I'd rather not have empty javadocs like this (ok to remove, no docs needed IMO, but if you want to keep it then please fill it in)

ywangd replied:
This is an oversight. Thanks for catching it. The javadoc should be in place now.

DaveCTurner commented on modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java:
```diff
@@ -326,29 +292,82 @@ ThreadPool getThreadPool() {
 
-        final AtomicLong putCount = new AtomicLong();
+    class StatsCollectors {
+        final ConcurrentMap<String, IgnoreNoResponseMetricsCollector> collectors;
```
No need for a concurrent map here, we populate it up front. An `EnumMap` would be best.

ywangd replied:
The stats are keyed by a pair of enums that is `Operation/Purpose`, e.g. `GetObject/ClusterState`. The two sets of Enums track different things. For example, the `BlobContainer.compareAndExchangeRegister()` method can be invoked for `Snapshot` purpose while the method internally uses both `ListObjects` and `PutMultipartObject` operations. It results into two entries in stats collection, i.e. `ListObjects/Snapshot` and `PutMultipartObject/Snapshot`. So it is not possible to have it as a EnumMap. 

We populate this map upfront mainly for BWC because existing API always reports all operations regardless whether the operation is actually used or not. Also for BWC reasons, we skip `Snapshot` in the stats key so that `ListObjects/Snapshot` is just `ListObjects`. The code initialise stats collections for purposes other than Snapshot on the fly. This also means they, e.g. `ListObjects/ClusterState` won't be initialised and not visible at all for Stateful.

I hope this makes sense.

DaveCTurner replied:
Ah sorry I see now, completely missed this the first time through.

DaveCTurner commented on server/src/main/java/org/elasticsearch/common/blobstore/support/FilterBlobContainer.java:
```diff
@@ -40,18 +41,18 @@ public BlobPath path() {
-    public boolean blobExists(String blobName) throws IOException {
-        return delegate.blobExists(blobName);
+    public boolean blobExists(OperationPurpose purpose, String blobName) throws IOException {
+        return delegate.blobExists(OperationPurpose.SNAPSHOT, blobName);
```
Throughout `FilterBlobContainer` we should pass the `purpose` argument rather than hard-coding it to `SNAPSHOT`.

ywangd replied:
I missed this file. Thanks for catching it. Your comment prompted me to do another sweep to check where the default value of `SNAPSHOT` is used. There are two scenarios where the default value is used:
1. Oversight like this one where this argument should be passed through from the caller method, also found in files such as FsBlobContainer etc. These are now fixed.
2. The actual call site or its approximate. An example is BlobStoreRepository or ChecksumBlobStoreFormat. These are the classes where we will leverage change of this PR to pass in the actual operation purpose. The equivalent on the stateless side is the ObjectStoreService. I think it makes sense to have them as follow-ups.

DaveCTurner commented on server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java:
```diff
@@ -117,7 +118,7 @@ public ChecksumBlobStoreFormat(
         throws IOException {
         String blobName = blobName(name);
-        try (InputStream in = blobContainer.readBlob(blobName)) {
+        try (InputStream in = blobContainer.readBlob(OperationPurpose.SNAPSHOT, blobName)) {
```
Hmm this purpose should come from the caller I think (can be a followup tho)

ywangd replied:
Yeah a follow up is better, see also [above](https://github.com/elastic/elasticsearch/pull/99615#discussion_r1337088235).

DaveCTurner commented on server/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java:
```diff
@@ -343,7 +344,13 @@ public void write(T obj, BlobContainer blobContainer, String name, boolean compr
         final String blobName = blobName(name);
-        blobContainer.writeMetadataBlob(blobName, false, false, out -> serialize(obj, blobName, compress, serializationParams, out));
+        blobContainer.writeMetadataBlob(
+            OperationPurpose.SNAPSHOT,
```
Likewise here, this purpose should come from the caller I think (can be a followup tho)

ywangd replied:
Yep same as above.

DaveCTurner commented on test/framework/src/main/java/org/elasticsearch/snapshots/mockstore/MockRepository.java:
```diff
@@ -640,21 +648,25 @@ private void beforeWrite(String blobName) throws IOException {
-            public void writeBlobAtomic(final String blobName, final BytesReference bytes, final boolean failIfAlreadyExists)
-                throws IOException {
+            public void writeBlobAtomic(
+                OperationPurpose purpose,
```
Nit: harmony

```suggestion
                final OperationPurpose purpose,
```

DaveCTurner commented on x-pack/plugin/snapshot-repo-test-kit/src/main/java/org/elasticsearch/repositories/blobstore/testkit/BlobAnalyzeAction.java:
```diff
@@ -341,17 +342,18 @@ public StreamInput streamInput() throws IOException {
                     if (atomic) {
                         try {
-                            blobContainer.writeBlobAtomic(request.blobName, bytesReference, failIfExists);
+                            blobContainer.writeBlobAtomic(OperationPurpose.SNAPSHOT, request.blobName, bytesReference, failIfExists);
```
We could reasonably distinguish the purpose of the repo analysis calls (in a followup tho)

DaveCTurner commented on modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java:
```diff
@@ -326,29 +292,82 @@ ThreadPool getThreadPool() {
+                key = operation.value;
+            } else {
+                key = operation.value + "/" + purpose.getValue();
+                if (false == collectors.containsKey(key)) {
```
Is this possible?

ywangd replied:
As commented [above](https://github.com/elastic/elasticsearch/pull/99615#discussion_r1337047607), this is to lazy initialize stats collections for purposes other than Snapshot, e.g. `ListObjects/ClusterState`.

DaveCTurner reviewed (commented):
I like it, and left fairly superficial comments.

Could you open a (draft) serverless PR which just updates the submodule to this branch to verify that it won't break the build there?

DaveCTurner commented on modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java:
```diff
@@ -326,29 +292,82 @@ ThreadPool getThreadPool() {
+            } else {
+                key = operation.value + "/" + purpose.getValue();
+                if (false == collectors.containsKey(key)) {
+                    collectors.put(key, buildMetricCollector(operation));
```
Suggest using `putIfAbsent` so that this step is atomic.

ywangd replied:
🤦 Right. Thanks!

henningandersen commented on server/src/main/java/org/elasticsearch/common/blobstore/OperationPurpose.java:
```diff
@@ -0,0 +1,26 @@
+        this.value = value;
+    }
+
+    public String getValue() {
```
I wonder about this name. I'd call it `getStatsKey` or just `getKey` rather than anything with `value`. Perhaps you have a specific motivation for the name though?

ywangd replied:
No specific reason. The word `value` is my default choice when I don't have a good name. I updated it to `getKey`.

ywangd committed:
Address feedback

henningandersen commented on modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java:
```diff
@@ -326,29 +292,82 @@ ThreadPool getThreadPool() {
+                Operation.PUT_MULTIPART_OBJECT,
+                Operation.DELETE_OBJECTS,
+                Operation.ABORT_MULTIPART_OBJECT
+            ).collect(Collectors.toConcurrentMap(ops -> ops.value, this::buildMetricCollector));
```
`toConcurrentMap` puzzled me. The javadoc says `There are no guarantees on the type, mutability, or serializability of the ConcurrentMap returned.`. I think we rely on it being mutable here. The fact that a potentially non-mutable concurrent map could be returned is weird though. Regardless, I'd prefer to collect in a map and then construct a concurrent map from it.

ywangd replied:
Thanks for the insights. You are right that the javadoc does suggest no such guarantee. I'll keep this in mind. For this PR, since I adopted your suggestion of *not* having the actual stats collection change, the concurrentMap is no longer needed. So it's now just a `Collectors.toUnmodifiableMap`.

henningandersen commented on modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java:
```diff
@@ -326,29 +292,82 @@ ThreadPool getThreadPool() {
+            if (purpose == OperationPurpose.SNAPSHOT) {
+                key = operation.value;
+            } else {
+                key = operation.value + "/" + purpose.getValue();
```
I think I'd prefer for the purpose to be a separate field in the resulting stats x-content. That would allow us to still look at write-object across all blobpurposes more easily. And could allow filtering on a purpose more easily.

I wonder if we should leave out the actual stats change from this PR in order to get the blob-purpose change in alone and be able to separate the review (the purpose change touches lots of trivial changes, whereas the stats change is a few more important changes).

ywangd replied:
> I wonder if we should leave out the actual stats change from this PR in order to get the blob-purpose change in alone and be able to separate the review

Good call. I think this simplify things for this PR since it is mostly superficial changes. I dropped relevant code changes as suggested.

henningandersen reviewed (commented):
This looks great. I have a few comments on naming. Also, I'd like to leave out the actual stats collection for now to separate these concerns (since I have comments on the structure there).

ywangd committed:
rename and remove changes to actual stats collection

ywangd committed:
Merge remote-tracking branch 'origin/main' into s3-more-granular-metrics

ywangd commented:
Thanks for the reviews. I have updated the PR accordingly. The most notable update is that I removed the changes to actual stats collection so that this PR focuses on adding the new purpose parameter (as suggested [here](https://github.com/elastic/elasticsearch/pull/99615#discussion_r1337086868)). This is a good call since I think we are not entirely aligned on the new stats structure. Deferring it gives us more time for discussion. Note that I still kept the refactoring for `getMetricsCollector` since it has no functional change and will be generally more useful in follow-ups.

ywangd committed:
fix method delegation

ywangd committed:
Merge remote-tracking branch 'origin/main' into s3-more-granular-metrics

ywangd commented:
I believe all reviews have been addressed. I'd appreciate if you could take another look. Thanks!

DaveCTurner commented on modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java:
```diff
@@ -74,14 +77,7 @@ class S3BlobStore implements BlobStore {
-    final RequestMetricCollector multiPartUploadMetricCollector;
-    final RequestMetricCollector deleteMetricCollector;
-    final RequestMetricCollector abortPartUploadMetricCollector;
+    private final StatsCollectors statsCollectors = new StatsCollectors();
```
Could we revert these changes to stats collection too? That way this PR is just about adding the `purpose` parameter to all the APIs, which is noisy but low-risk, and follow-ups which change the behaviour like here will be easier to review and will let us pin down any later problems with `git bisect`.

ywangd replied:
Sure I reverted these changes as well. I believe this PR is now a pure refactor. Hence I adjusted the labels accordingly as well.

DaveCTurner commented on server/src/main/java/org/elasticsearch/common/blobstore/BlobPurpose.java:
```diff
@@ -0,0 +1,26 @@
+
+package org.elasticsearch.common.blobstore;
+
+public enum BlobPurpose {
```
Sorry for nitpicking about naming here but I don't like the new name `BlobPurpose`. We're not distinguishing _blobs_ by purpose, it's the _operations_ on them.

ywangd replied:
OK I reverted it back to `OperationPurpose`.

DaveCTurner reviewed (commented):
I left a couple more (somewhat superficial) comments.

ywangd committed:
Merge remote-tracking branch 'origin/main' into s3-more-granular-metrics

ywangd committed:
more revert to address feedback

ywangd committed:
Delete docs/changelog/99615.yaml

DaveCTurner reviewed (approved):
LGTM thanks for the extra iterations Yang

henningandersen commented on modules/repository-s3/src/internalClusterTest/java/org/elasticsearch/repositories/s3/S3BlobStoreRepositoryTests.java:
```diff
@@ -202,6 +206,31 @@ public void testAbortRequestStats() throws Exception {
+
+        final BlobPath blobPath = repository.basePath().add(randomAlphaOfLength(10));
+        final BlobContainer blobContainer = blobStore.blobContainer(blobPath);
+        final OperationPurpose purpose = randomValueOtherThan(OperationPurpose.SNAPSHOT, () -> randomFrom(OperationPurpose.values()));
```
I wonder why we exclude `SNAPSHOT` here? I'd expect the test to also succeed for that and furthermore for the validation of SNAPSHOT to be equally valid?

ywangd replied:
This test had two parts where the 1st part uses `SNAPSHOT` and the 2nd part randomizes from purposes other than SNAPSHOT. Since this PR is now a pure refactor, the test mostly lost its purpose. Now it really just shows the refactor has no impact on output. In this case, you are right that there is no need to exclude SNAPSHOT. I have updated accordingly.

henningandersen commented on modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java:
```diff
@@ -243,9 +244,14 @@ public void deleteBlobsIgnoringIfNotExists(Iterator<String> blobNames) throws IO
     }
 
-    private void deletePartition(AmazonS3Reference clientReference, List<String> partition, AtomicReference<Exception> aex) {
+    private void deletePartition(
```
nit: all other places have `purpose` as first arg, I'd like to keep it consistent, so can we move it to be first here too?

ywangd replied:
Thanks for catching it. It was my intention to keep it consistently but apparently missed this one and the one below. Corrected now.

henningandersen commented on modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java:
```diff
@@ -264,7 +270,7 @@ private void deletePartition(AmazonS3Reference clientReference, List<String> par
     }
 
-    private static DeleteObjectsRequest bulkDelete(S3BlobStore blobStore, List<String> blobs) {
+    private static DeleteObjectsRequest bulkDelete(S3BlobStore blobStore, List<String> blobs, OperationPurpose purpose) {
```
nit: all other places have `purpose` as first arg, I'd like to keep it consistent, so can we move it to be first here too?

henningandersen commented on modules/repository-url/src/test/java/org/elasticsearch/common/blobstore/url/HttpURLBlobStoreTests.java:
```diff
@@ -126,8 +127,14 @@ String getBlobName() {
-        expectThrows(IOException.class, () -> container.readBlob(blobName, content.length + 1, content.length).read());
+        expectThrows(
+            IllegalArgumentException.class,
+            () -> container.readBlob(OperationPurpose.SNAPSHOT, blobName, -1, content.length).read()
```
It would be nice to randomize the purpose here (and in many places elsewhere). I think it can be a follow-up though.

ywangd replied:
Yep we should have follow-ups for actual usages of the purpose parameter. In this PR, all callsite simply uses the SNAPSHOT as the default.

henningandersen commented on server/src/main/java/org/elasticsearch/common/blobstore/BlobContainer.java:
```diff
@@ -37,39 +37,55 @@ public interface BlobContainer {
-     * @param   blobName
-     *          The name of the blob whose existence is to be determined.
-     * @return  {@code true} if a blob exists in the {@link BlobContainer} with the given name, and {@code false} otherwise.
+     * @param purpose The purpose of the operation, useful for stats collection.
```
I think there are several potential uses of this, stats collection, storage classes, priority. Perhaps we can move the examples to `OperationPurpose` javadoc and leave the example out here (and in other methods in this class)?

```suggestion
     * @param purpose The purpose of the operation
```

ywangd replied:
Makes sense. I removed it from individual method and added it (with a bit more explanation) to the Enum class. Btw, the other usages sound pretty exciting!

henningandersen commented on server/src/main/java/org/elasticsearch/common/blobstore/OperationPurpose.java:
```diff
@@ -0,0 +1,26 @@
+
+package org.elasticsearch.common.blobstore;
+
+public enum OperationPurpose {
```
I see us mentioning stats collection in every method javadoc where this is passed. I think that is only one use out of potentially several. Instead, we could supply that information here and leave the method level javadoc without the use case.

henningandersen reviewed (approved):
LGTM.

Left a few minor comments but no need for another round. Thanks for the extra iterations on this.

ywangd committed:
address feedback

ywangd committed:
fix compilation
